move analysis code in repo

bloodearnest opened the pull request:
feat: add the codelist action code into the repo
   
This creates self contained action code in `actions/codelist`. In aims
to consolidate all the codelist action specific information in once
place. The goal for this is primarily to local all the related code in
once location, but also to support multiple OSI actions in the future.
    
As part of this, it cleans up the file generation, by getting rid of the
the awkward variables.py set up, and moving to using cli args in the
project.yaml
    
It adds a full integration test, which uses opensafely to run the
action. This is unfortunately very messy, as we have no
expectations/dummy data options for the `generate_codelist_report`
action. This requires us to fake it, which is ugly as sin. But it does
work.

bloodearnest commented:
Ah, I hadn't seen them, will do!

lucyb commented on tests/factories.py:
```diff
@@ -20,7 +20,7 @@ class Meta:
     title = factory.Sequence(lambda n: f"Analysis Request {n}")
-    start_date = factory.Faker("date")
-    end_date = factory.Faker("date")
+    start_date = "2021-01-01"
```
How come the date has been hard-coded to a single day?

bloodearnest replied:
Because we need something valid as a default. The Faker versions would sometime produce a an end date before a start date, resulting in no outputs

It's not hardcoded, its just the default. You can do `AnalysisRequestFactory(start_date=...)` to alter it.

I guess I could use the current default [START_DATE](https://github.com/opensafely-core/interactive.opensafely.org/blob/main/interactive/models.py#L35)?

lucyb replied:
Fair point about the end date being before the start date. I guess I was a bit worried about only querying one day. The actual application code is using START_DATE and END_DATE, so maybe the tests should use that too?

lucyb commented:
What's the thinking behind putting things into `actions/codelist/analysis`? Why not put everything into the top-level actions dir?

I know we've used the name genearte_codelist_report in the cohortextractor, but I don't know if we should use the same terminology in OSI (i.e the `codelist` dir), since we're likely to expand the report in future

bloodearnest commented:
> What's the thinking behind putting things into `actions/codelist/analysis`? Why not put everything into the top-level actions dir?
> 
> I know we've used the name genearte_codelist_report in the cohortextractor, but I don't know if we should use the same terminology in OSI (i.e the `codelist` dir), since we're likely to expand the report in future

The thinking was that we'd support multiple OSI "actions" in future, the first one being a single codelist report, hence `codelist`. So everything for this action is in the `actions/codelist` dir. Once inside that dir, it should look and feel like normal study code, hence `./analysis`, ./output`, etc. I have a follow on PR already done that makes these "actions" easy to work on for researchers, making them just like a regular local study development. I had wondered about moving the tests in there too, so `actions/codelist/tests`, for even better local development of the study code.

Maybe `actions` is the wrong noun, as it's actually a study with multiple actions. But `reports` is kinda already taken, although arguably OSI is the future of reports site anyway.

bloodearnest commented:
I'm not sure about the nameing or terminology, but I'm fairly sure that splitting out current report into its own subdir, to make it easy to support multiple kinds of OSI reports in future, is a good idea.

lucyb commented:
> Maybe actions is the wrong noun, as it's actually a study with multiple actions. But reports is kinda already taken, although arguably OSI is the future of reports site anyway.

So far OSI has used "analysis" or "report", so maybe using reports makes sense? It's an OSI Report, rather than a OS Report. I'm sure that won't be confusing.. maybe? Naming things is hard.

Having the `analysis` subdir within `codelist` like the other studies makes sense now. I wonder if it's worth moving the tests in and having a justfile in there (like we do for the docker dir) would be helpful for Louis or others working on that code?

lucyb commented:
> I'm not sure about the nameing or terminology, but I'm fairly sure that splitting out current report into its own subdir, to make it easy to support multiple kinds of OSI reports in future, is a good idea.

I think that makes sense. At some point we'll have to think about how the reports will be versioned and evolve. Maybe we'll only ever have one type? Seb's comment about being able to consume any reusable action is an interesting one, not sure how that changes things.

lucyb commented:
I tried running the tests locally and got a failure on the third time I ran them (they were slow, so I was interested in that). Seems like a non-deterministic hypothesis thing?

```
============================================================= short test summary info =============================================================
FAILED tests/actions/test_codelist.py::test_group_low_values - AssertionError: assert ['0', '1', 'code'] == ['0', '1']
====================================== 1 failed, 12 passed, 64 deselected, 206 warnings in 123.68s (0:02:03) ======================================
error: Recipe `test` failed on line 115 with exit code 1
```

bloodearnest commented:
> > I'm not sure about the nameing or terminology, but I'm fairly sure that splitting out current report into its own subdir, to make it easy to support multiple kinds of OSI reports in future, is a good idea.
> 
> I think that makes sense. At some point we'll have to think about how the reports will be versioned and evolve. Maybe we'll only ever have one type? Seb's comment about being able to consume any reusable action is an interesting one, not sure how that changes things.

Yeah, we'll still need to wrap that reusable action in templated project.yaml, even if there's only once job to run.

Regards versioning, this doesn't help us that much - we'd only be able to use the currently deployed version of a report, unless we have a parallel `codelist_dev` report?

lucyb commented:
> If it's ok with you, I will do that in my already par-baked follow-on PR that focuses on development experience for researchers?

Sounds like an excellent plan

lucyb commented:
> Regards versioning, this doesn't help us that much - we'd only be able to use the currently deployed version of a report, unless we have a parallel codelist_dev report?

I suggest we say that's out of scope for now!

lucyb commented on actions/codelist/__init__.py:
```diff
@@ -0,0 +1,99 @@
+low_count_threshold = 100
+rounding_base = 10
+
+PROJECT_YAML = """
```
Is this being removed from submit.py in a future PR?

lucyb commented on actions/codelist/analysis/generate_measures.py:
```diff
@@ -0,0 +1,62 @@
+import numpy as np
```
This module was refactored around a month ago. Would be good to use the latest version of the code, especially because it was changed to be more testable.

bloodearnest replied:
I copied the code from interactive repo. Was there somewhere else?

bloodearnest replied:
Hmm, it seems test-point-and-click and interactive diverge quite a bit

lucyb replied:
Once your change is in and being used, then we should archive the test-point-and-click repo to save confusion

lucyb commented on actions/codelist/analysis/weekly_summary.py:
```diff
@@ -0,0 +1,31 @@
+# from study_utils import create_top_5_code_table
```
This seems to be a new module that isn't used anywhere. What's its purpose?

bloodearnest replied:
I think I've made a snafu and based this work on an older version of the report code :facepalm:

bloodearnest replied:
removed. It's in the interactive code atm, but its not used

bloodearnest commented:
Right, I had made a mess of copying the current report code in. I had an out-of-date version of the interactive repo that I used. Fixed, and cleaner for it, though it underscores we should test this before switching it on in prod.

I've also renamed `actions` to `reports` everywhere.

bloodearnest commented:
```diff -u ../interactive/analsysi reports/codelist/analysis
Only in ../interactive/analysis/: generate_dummy_data.py
diff -u ../interactive/analysis/generate_measures.py reports/codelist/analysis/generate_measures.py
--- ../interactive/analysis/generate_measures.py	2022-06-17 17:34:02.730106882 +0100
+++ reports/codelist/analysis/generate_measures.py	2022-06-17 17:41:43.531450089 +0100
@@ -1,19 +1,17 @@
-import sys
-
-import pandas as pd
 import numpy as np
+import pandas as pd
 from study_utils import (
     calculate_rate,
-    redact_events_table,
     convert_weekly_to_monthly,
+    redact_events_table,
     round_values,
 )
-from variables import low_count_threshold, rounding_base, rounding_base_practice_count
-
-if len(sys.argv) > 1:
-    output_dir = sys.argv[1]
-else:
-    output_dir = "output"
+from variables import (
+    low_count_threshold,
+    output_dir,
+    rounding_base,
+    rounding_base_practice_count,
+)
 
 
 def practice_counts(counts_table, list_sizes):
Only in reports/codelist/analysis/: __init__.py
Only in reports/codelist/analysis/: __pycache__
diff -u ../interactive/analysis/study_utils.py reports/codelist/analysis/study_utils.py
--- ../interactive/analysis/study_utils.py	2022-06-17 17:33:04.051063763 +0100
+++ reports/codelist/analysis/study_utils.py	2022-06-17 17:42:14.008472990 +0100
@@ -1,62 +1,5 @@
-from datetime import date, datetime
 import numpy as np
-from dateutil import relativedelta
-from cohortextractor import (
-    StudyDefinition,
-    codelist,
-    patients,
-)
-
-from variables import study_start_date, study_end_date
-
-
-def build_study_definition_for_counts(codelist):
-    start_date = datetime.strptime(study_start_date, "%Y-%m-%d").date()
-    end_date = datetime.strptime(study_end_date, "%Y-%m-%d").date()
-
-    return StudyDefinition(
-        default_expectations={
-            "date": {"earliest": study_start_date, "latest": study_end_date},
-            "rate": "uniform",
-            "incidence": 0.5,
-        },
-        index_date=study_start_date,
-        population=patients.satisfying(
-            "currently_registered OR has_died",
-            currently_registered=patients.registered_as_of(study_end_date),
-            has_died=patients.with_death_recorded_in_primary_care(
-                between=[study_start_date, study_end_date], returning="binary_flag"
-            ),
-        ),
-        **calculate_code_frequency(start_date, end_date, codelist),
-    )
-
-
-def calculate_code_frequency(start_date, end_date, selected_codes):
-    start_date_formatted = date.strftime(start_date, "%Y-%m-%d")
-    end_date_formatted = date.strftime(end_date, "%Y-%m-%d")
-
-    variables = {}
-    for code in selected_codes:
-        variables[f"code_{code}"] = patients.with_these_clinical_events(
-            codelist([code], system="snomed"),
-            between=[start_date_formatted, end_date_formatted],
-            episode_defined_as="series of events each <= 0 days apart",
-            returning="number_of_episodes",
-            return_expectations={
-                "incidence": 0.1,
-                "int": {"distribution": "normal", "mean": 3, "stddev": 1},
-            },
-        )
-    return variables
-
-
-def last_day_of_month(dt):
-    return dt + relativedelta.relativedelta(day=31)
-
-
-def last_day_of_week(dt):
-    return dt + relativedelta.relativedelta(day=7)
+import pandas as pd
 
 
 def group_low_values(df, count_column, code_column, threshold):
@@ -74,28 +17,34 @@
 
     # get sum of any values <= threshold
     suppressed_count = df.loc[df[count_column] <= threshold, count_column].sum()
+    suppressed_df = df.loc[df[count_column] > threshold, count_column]
 
-    # if no values are suppressed we don't need to do anything
-    if suppressed_count == 0:
-        pass
+    # if suppressed values >0 ensure total suppressed count > threshold.
+    # Also suppress if all values 0
+    if (suppressed_count > 0) | (
+        (suppressed_count == 0) & (len(suppressed_df) != len(df))
+    ):
 
-    # if suppressed values >0 ensure total suppressed count > threshold
-    else:
         # redact counts <= threshold
         df.loc[df[count_column] <= threshold, count_column] = np.nan
 
-        # if suppressed count <= threshold redact further values
-        while suppressed_count <= threshold:
-            suppressed_count += df[count_column].min()
-            df.loc[df[count_column].idxmin(), :] = np.nan
+        # If all values 0, suppress them
+        if suppressed_count == 0:
+            df.loc[df[count_column] == 0, :] = np.nan
+
+        else:
+            # if suppressed count <= threshold redact further values
+            while suppressed_count <= threshold:
+                suppressed_count += df[count_column].min()
+                df.loc[df[count_column].idxmin(), :] = np.nan
 
         # drop all rows where count column is null
         df = df.loc[df[count_column].notnull(), :]
 
         # add suppressed count as "Other" row (if > threshold)
         if suppressed_count > threshold:
-            suppressed_count = {code_column: "Other", count_column: suppressed_count}
-            df = df.append(suppressed_count, ignore_index=True)
+            suppressed_count = {"code": "Other", count_column: suppressed_count}
+            df = pd.concat([df, pd.DataFrame([suppressed_count])], ignore_index=True)
 
     return df
 
@@ -184,10 +133,12 @@
 
 
 def round_values(x, base=5):
-    if x is not None and not str(x).startswith("<=") and not np.isnan(x):
-        rounded = int(base * round(x / base))
-    else:
-        rounded = np.nan
+    rounded = x
+    if isinstance(x, (int, float)):
+        if np.isnan(x):
+            rounded = np.nan
+        else:
+            rounded = int(base * round(x / base))
     return rounded
 
 
@@ -216,7 +167,9 @@
     num_dates_over = num_dates % 4
     if num_dates_over != 0:
         # drop rows from counts table
-        counts_table = counts_table.loc[~counts_table["date"].isin(dates[0:num_dates_over]), counts_table.columns]
+        counts_table = counts_table.loc[
+            ~counts_table["date"].isin(dates[0:num_dates_over]), counts_table.columns
+        ]
 
         # drop dates from dates list
         dates = dates[num_dates_over:]
@@ -227,7 +180,9 @@
         date_group = dates[i : i + 4]
         for date in date_group:
             dates_map[date] = date_group[0]
-    counts_table.loc[counts_table.index,"date"] = counts_table.loc[counts_table.index,"date"].map(dates_map)
+    counts_table.loc[counts_table.index, "date"] = counts_table.loc[
+        counts_table.index, "date"
+    ].map(dates_map)
 
     # group into 4 weeks
     counts_table = (
Only in ../interactive/analysis/: tests
diff -u ../interactive/analysis/top_codes_table.py reports/codelist/analysis/top_codes_table.py
--- ../interactive/analysis/top_codes_table.py	2022-05-13 14:47:12.757661790 +0100
+++ reports/codelist/analysis/top_codes_table.py	2022-06-17 12:00:16.261792239 +0100
@@ -1,14 +1,6 @@
-import sys
-
 import pandas as pd
 from study_utils import create_top_5_code_table
-from variables import low_count_threshold, rounding_base
-
-
-if len(sys.argv) > 1:
-    output_dir = sys.argv[1]
-else:
-    output_dir = "output"
+from variables import low_count_threshold, output_dir, rounding_base
 
 
 def main():
diff -u ../interactive/analysis/variables.py reports/codelist/analysis/variables.py
--- ../interactive/analysis/variables.py	2022-06-17 17:33:04.052063764 +0100
+++ reports/codelist/analysis/variables.py	2022-06-17 17:34:38.020132897 +0100
@@ -1,5 +1,19 @@
-study_start_date = "2021-01-01"
-study_end_date = "2021-12-31"
-low_count_threshold = 100
-rounding_base = 10
-rounding_base_practice_count = 5
+import argparse
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--start-date")
+parser.add_argument("--end-date")
+parser.add_argument("--output-dir")
+parser.add_argument("--low-count-threshold", default=100, type=int)
+parser.add_argument("--rounding-base", default=10, type=int)
+parser.add_argument("--rounding-base-practice-count", default=5, type=int)
+
+args = parser.parse_args()
+
+# minimum viable conversion from cli args to module variables
+locals().update(vars(args))
+
+
+if __name__ == "__main__":
+    print(args)
Only in ../interactive/analysis/: weekly_summary.py
```

LFISHER7 reviewed (approved):
I've added one small comment but otherwise all the analysis code looks good to me

bloodearnest commented on reports/codelist/analysis/study_utils.py:
```diff
@@ -0,0 +1,194 @@
+        df.loc[df[count_column] <= threshold, count_column] = np.nan
+
+        # If all values 0, suppress them
+        if suppressed_count == 0:
+            df.loc[df[count_column] == 0, :] = np.nan
```
I'm gonna merge as is, we can clean things up later
